dev/env/scripts/lib.sh

```diff
@@ -0,0 +1,191 @@
+    export IMAGE_PULL_DOCKER_CONFIG=${IMAGE_PULL_DOCKER_CONFIG:-$IMAGE_PULL_DOCKER_CONFIG_DEFAULT}
+    export KUBECONF_CLUSTER_SERVER_OVERRIDE=${KUBECONF_CLUSTER_SERVER_OVERRIDE:-$KUBECONF_CLUSTER_SERVER_OVERRIDE_DEFAULT}
+    export INHERIT_IMAGEPULLSECRETS=${INHERIT_IMAGEPULLSECRETS:-$INHERIT_IMAGEPULLSECRETS_DEFAULT}
+    export SPAWN_LOGGER=${SPAWN_LOGGER:-$SPAWN_LOGGER_DEFAULT}
+    export DUMP_LOGS=${DUMP_LOGS:-$DUMP_LOGS_DEFAULT}
+    export OPERATOR_SOURCE=${OPERATOR_SOURCE:-$OPERATOR_SOURCE_DEFAULT}
+    export INSTALL_OLM=${INSTALL_OLM:-$INSTALL_OLM_DEFAULT}
+    export ENABLE_DB_PORT_FORWARDING=${ENABLE_DB_PORT_FORWARDING:-$ENABLE_DB_PORT_FORWARDING_DEFAULT}
+    export ENABLE_FM_PORT_FORWARDING=${ENABLE_FM_PORT_FORWARDING:-$ENABLE_FM_PORT_FORWARDING_DEFAULT}
+    export AUTH_TYPE=${AUTH_TYPE:-$AUTH_TYPE_DEFAULT}
+    export FINAL_TEAR_DOWN=${FINAL_TEAR_DOWN:-$FINAL_TEAR_DOWN_DEFAULT}
+
+    export FLEET_MANAGER_IMAGE="${FLEET_MANAGER_IMAGE:-$FLEET_MANAGER_IMAGE_DEFAULT}"
+    # When transferring images without repository hostname to Minikube it gets prefixed with "docker.io" automatically.
+    if [[ "$FLEET_MANAGER_IMAGE" =~ ^fleet-manager-.*/fleet-manager:.* ]]; then
+        export FULL_FLEET_MANAGER_IMAGE="docker.io/${FLEET_MANAGER_IMAGE}"
+    else
+        export FULL_FLEET_MANAGER_IMAGE="${FLEET_MANAGER_IMAGE}"
+    fi
+
+    verify_environment
+
+    disable_debugging
+    enable_debugging_if_necessary
+}
+
+disable_debugging() {
+    if [[ "$DEBUG" != "trace" ]]; then
+        set +x
+    fi
+}
+
+enable_debugging_if_necessary() {
+    if [[ "$DEBUG" != "false" ]]; then
+        set -x
+    fi
+}
+
+wait_for_container_to_appear() {
+    local namespace="$1"
+    local pod_selector="$2"
+    local container_name="$3"
+    local status=$($KUBECTL -n "$ACSMS_NAMESPACE" get pod -l "${pod_selector}" -o jsonpath="{.items[0].status.initContainerStatuses[?(@.name == '${container_name}')]} {.items[0].status.containerStatuses[?(@.name == '${container_name}')]}")
+    local state=$(echo "${status}" | jq -r ".state | keys[]")
+    log "Waiting for container ${container_name} within pod ${pod_selector} in namespace ${namespace} to appear..."
+    for i in $(seq 10); do
+        if [[ "$state" == "terminated" || "$state" == "running" ]]; then
+            echo "Container ${container_name} is in state ${state}"
+            break
+        fi
+        sleep 1
+    done
+}
+
+wait_for_container_to_become_ready() {
+    local namespace="$1"
+    local pod_selector="$2"
+
+    log "Waiting for pod ${pod_selector} within namespace ${namespace} to become ready..."
+
+    for i in $(seq 10); do
+        if $KUBECTL -n "$ACSMS_NAMESPACE" wait --timeout=5s --for=condition=ready pod -l "$pod_selector" 2>/dev/null >&2; then
+            break
+        fi
+        sleep 1
+    done
+    $KUBECTL -n "$ACSMS_NAMESPACE" wait --timeout=20s --for=condition=ready pod -l "$pod_selector"
```

Comment:
I'm confused about the different waiting mechanisms here. Why the loop and why running the same kubectl wait command again after the loop? Wouldn't a single wait with higher timeout be enough?
```suggestion
    $KUBECTL -n "$ACSMS_NAMESPACE" wait --timeout=60s --for=condition=ready pod -l "$pod_selector"
```

Reply:
I see this has changed, but what I meant was that we could remove the for loop completely, now we have both. Is there a reason for this, that I'm not aware of?

Reply:
I can explain. Again, I agree this seems weird, but I found this solution to be the most reliable from the solutions I have tried.

The problem is that the pods do not exist initially, therefore a plain `kubectl wait` can in fact fail immediately (hence terminating the script). This is of course not what we want.

I *could* just add something like `sleep 10` (and guessing a reasonable delay value), but what I am trying here is a little bit smarter: I try multiple times (10) to wait for a short amount of time. When the pod is not there, the wait will fail, I will sleep and try again to wait. Also, if the pod is already there, I don't want to wait for too long, because the *actual* kubectl wait comes after to loop. 

Probably we can modify this so that we do it in two *different* steps. First we wait for the pod to come up (without using `kubectl wait`) and then we do the `kubectl wait`.

I would suggest that I look into that in my folllow-up polishing PR, ok?

Reply:
Ok, got it now, thanks for explaining.